bauble/src/convert.rs

```diff
@@ -0,0 +1,359 @@
+use crate::{
+    parse::{Ident, Path},
+    spanned::{Span, Spanned},
+    value::{ConvertionError, TypeInfo},
+    Attributes, Val, Value, ValueKind,
+};
+use num_traits::ToPrimitive;
+#[derive(Debug, PartialEq)]
+pub enum VariantKind {
+    Struct,
+    Tuple,
+    Path,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum DeserializeError {
+    WrongTypePath {
+        expected: TypeInfo,
+        found: Spanned<TypeInfo>,
+    },
+    MissingTypePath {
+        ty: TypeInfo,
+        span: Span,
+    },
+    MissingField {
+        field: String,
+        ty: TypeInfo,
+        span: Span,
+    },
+    UnexpectedField {
+        field: Ident,
+        ty: TypeInfo,
+    },
+    WrongTupleLength {
+        expected: usize,
+        found: usize,
+        ty: TypeInfo,
+        span: Span,
+    },
+    WrongArrayLength {
+        expected: usize,
+        found: usize,
+        ty: TypeInfo,
+        span: Span,
+    },
+    UnknownVariant {
+        variant: Spanned<String>,
+        kind: VariantKind,
+        ty: TypeInfo,
+    },
+    WrongFields {
+        found: VariantKind,
+        expected: VariantKind,
+        ty: TypeInfo,
+    },
+    MissingVariantName {
+        ty: TypeInfo,
+        span: Span,
+    },
+    NotAVariant {
+        ty: TypeInfo,
+        path: Spanned<Path>,
+    },
+    UnexpectedAttribute {
+        attribute: Ident,
+        ty: TypeInfo,
+    },
+    WrongKind {
+        expected: ValueKind,
+        found: ValueKind,
+        ty: TypeInfo,
+        span: Span,
+    },
+    WrongKindForId {
+        expected: String,
+        found: ValueKind,
+        ty: String,
+        span: Span,
+    },
+    Custom {
+        message: String,
+        span: Span,
+    },
+    Convertion(Spanned<ConvertionError>),
+}
+
+// TODO Maybe `unsafe trait`?
+pub trait BaubleAllocator<'a> {
+    type Out<T>
+    where
+        Self: 'a;
+
+    /// # Safety
+    /// Allocations in `value` have to be allocated with the allocator from `allocator`
+    unsafe fn wrap<T>(&'a self, value: T) -> Self::Out<T>;
+    /// # Safety
+    /// If validated an item must be placed within the same allocator.
+    unsafe fn validate<T>(&'a self, value: Self::Out<T>) -> Result<T, Box<DeserializeError>>;
+}
+
+pub struct DefaultAllocator;
+
+impl<'a> BaubleAllocator<'a> for DefaultAllocator {
+    type Out<T> = T;
+
+    unsafe fn wrap<T>(&'a self, value: T) -> Self::Out<T> {
+        value
+    }
+
+    unsafe fn validate<T>(&'a self, value: Self::Out<T>) -> Result<T, Box<DeserializeError>> {
+        Ok(value)
+    }
+}
+
+pub trait FromBauble<'a, A: BaubleAllocator<'a> = DefaultAllocator>: Sized {
+    fn from_bauble(data: Val, allocator: &'a A) -> Result<A::Out<Self>, Box<DeserializeError>>;
+}
+
+impl<'a, T: FromBauble<'a>> FromBauble<'a> for Vec<T> {
+    fn from_bauble(
+        Val {
```

Comment:
Wow, I did NOT know you could do destructuring in function arguments like that! Pretty cool!